_drafts/firmware-shell.md

```diff
@@ -0,0 +1,992 @@
+}
+```
+
+`cli_cmd_hello` outputs its response through a function called `shell_put_line`,
+which we convert into a mock for the purposes of testing.
+
+We can now run our simple test and verify it works!
+
+```
+$ cd complex/tests
+
+$ make
+/usr/bin/make -f interrupt/example/firmware-shell/complex/tests/makefiles/Makefile_test_shell_commands.mk
+compiling test_shell_commands.cpp
+compiling AllTests.cpp
+compiling shell_commands.c
+Building archive build/test_shell_commands/lib/libtest_shell_commands.a
+a - build/test_shell_commands/objs/interrupt/example/firmware-shell/complex/shell_commands.o
+Linking build/test_shell_commands/test_shell_commands_tests
+Running build/test_shell_commands/test_shell_commands_tests
+.
+OK (1 tests, 1 ran, 1 checks, 0 ignored, 0 filtered out, 0 ms)
+```
+
+For more reference, it's worth checking out how the
+[Memfault Firmware SDK](https://github.com/memfault/memfault-firmware-sdk/blob/master/tests/src/test_memfault_demo_cli.cpp)
+and
+[Skip Scooter's Anchor](https://github.com/rideskip/anchor/tree/master/console/tests)
+libraries test their command-line implementations.
+
+## Tips and Considerations
+
+### Formatted Output
+
+There are two consumers of a device's shell: a developer and a machine hooked up
+to the device through serial. Therefore, an output format should be both human
+and machine-readable.
+
+I suggest that all commands succeed or fail in the same way, either printing
+`OK` or `FAIL,<code>`. It might help to build helper functions that do this for
+you:
+
+```c
+void shell_result_success(void) {
+  shell_put_line("> OK");
+}
+
+void shell_result_fail(int code) {
+  char buf[32];
+  snprint(buf, sizeof(buf), "> FAIL,%d", code);
+  shell_put_line(buf);
+}
+
+```
+
+One common format that is used in CSV, which is mostly readable depending on
+what is being output. Below, we show an example of a `dump_heap` command which
+would spit out a CSV, which a developer could then copy into Excel or Google
+Sheets and do further inspections.
+
+```
+$ dump_heap
+addr,size
+0x00200000,64
+0x00200040,128
+...
+```
+
+If you are trying to output files or raw binary buffers into the shell, I
+recommend using base64 or hexdump encoding, which will make it easier for
+automated systems to ingest and convert the output.
+
+### Extra Shell Features
+
+Every developer loves a full-featured shell. These features might include being
+able to backspace and delete keys, use the left and right arrow keys for
+in-place editing, Emacs shortcuts, command history and completion, argument
+parsing, and help menus.
+
+To learn how to add some of these features to your project's shell, it's best to
+look at a few open-source projects. Unfortunately, there aren't many _great_
+shell's publicly available for tiny firmware.
+
+- [Zephyr Shell](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/shell)
+- [Skip Scooters Anchor Console](https://github.com/rideskip/anchor/tree/master/console)
+
```

Comment:
The esp-idf shell does some neat stuff too. A lot of it is offloaded to the desktop:  https://github.com/espressif/esp-idf/blob/master/tools/idf_monitor.py#L3-L10

console source itself is here https://github.com/espressif/esp-idf/tree/master/components/console